### PR TITLE
Move dapp connection logic to EmbarkJS; put Contract inside Blockchain Component

### PR DESCRIPTION
**Pull request by iurimatias**

### CODE REVIEW

File: src/blockchain.js [hunk 1]
```diff
@@ -1,3 +1,131 @@
...
+      return null;
+    });
+    return;
+  }
+  web3.eth.getAccounts(cb);
+};
+
+let Blockchain = {};
+
+Blockchain.connect = function(connectionList, opts, doneCb) {
```
Comment (michaelsbradleyjr) [inline src/blockchain.js:20]: Might it make sense to de-globalize W/web3 and have it be an argument to `Blockchain.connect`?

**Review by alaibe — APPROVED** [review]

File: src/blockchain.js [hunk 2]
```diff
@@ -1,3 +1,131 @@
...
+      iteratee = memo;
+      memo = [];
+    } else {
+      throw new TypeError('expected callback to be a function');
+    }
+  }
+
+  if (!Array.isArray(arr)) {
+    cb(new TypeError('expected an array'));
+    return;
```
Comment (jrainville) [inline src/blockchain.js]: You can put all the `return callback()` on the same line

File: src/blockchain.js [hunk 3]
```diff
@@ -1,3 +1,131 @@
+
+function reduce(arr, memo, iteratee, cb) {
```
Comment (jrainville) [inline src/blockchain.js]: You know that ES6 has it's own Array.prototype.reduce` that does pretty much the same thing, just without the callback.
Eg: https://github.com/embark-framework/embark/blob/develop/lib/utils/utils.js#L422

File: src/blockchain.js [hunk 4]
```diff
@@ -1,3 +1,131 @@
...
+        } else {
+          next(null, false)
+        }
+      });
+    }, function(err, _result) {
+      self.web3 = web3;
+      getAccounts(function(err, accounts) {
+        if (opts.warnAboutMetamask) {
+          if (web3.eth.currentProvider && web3.eth.currentProvider.isMetaMask) {
+            console.log("%cNote: Embark has detected you are in the development environment and using Metamask, please make sure Metamask is connected to your local node", "font-size: 2em");
```
Comment (jrainville) [inline src/blockchain.js]: Maybe use `.warn`

File: src/blockchain.js [hunk 5]
```diff
@@ -1,3 +1,131 @@
...
+        if (value.indexOf('ws://') >= 0) {
+          web3.setProvider(new Web3.providers.WebsocketProvider(value));
+        } else {
+          web3.setProvider(new Web3.providers.HttpProvider(value));
+        }
+      } else if (value === '$WEB3') {
+        return next(null, '');
+      }
+
+      getAccounts(function(err, account) {
```
Comment (jrainville) [inline src/blockchain.js:41]: Why do you getAccounts here, if you do it in the next function anyway? Maybe I just don't understand the function flow.

File: src/blockchain.js [hunk 6]
```diff
@@ -1,3 +1,131 @@
...
+        if (opts.warnAboutMetamask) {
+          if (web3.eth.currentProvider && web3.eth.currentProvider.isMetaMask) {
+            console.log("%cNote: Embark has detected you are in the development environment and using Metamask, please make sure Metamask is connected to your local node", "font-size: 2em");
+          }
+        }
+        if (accounts) {
+          web3.eth.defaultAccount = accounts[0];
+        }
+        cb();
+        doneCb(err);
```
Comment (jrainville) [inline src/blockchain.js:56]: Two callbacks?

File: src/index.js [hunk 7]
```diff
@@ -1,24 +1,29 @@
...
 };
 
-EmbarkJS.Contract = Contract;
+EmbarkJS.Blockchain = Blockchain;
 EmbarkJS.Storage = Storage;
 EmbarkJS.Names = Names;
 EmbarkJS.Messages = Messages;
 EmbarkJS.Utils = Utils;
+EmbarkJS.Contract = function() {
+  console.error("deprecated: please use EmbarkJS.Blockchain.Contract instead");
```
Comment (jrainville) [inline src/index.js]: Should probably throw an error instead. Because it will throw anyway at some place else

File: src/blockchain.js (hunk 1 above)
Comment (iurimatias) [reply src/blockchain.js:20]: I agree it's a good idea, however I rather do it in a different PR (possibly one that moves out the web3.js implementation itself) given the possible breaking changes this one already risks

File: src/blockchain.js (hunk 5 above)
Comment (iurimatias) [reply src/blockchain.js:41]: basically it's a way to verify the connection status... it's stupid but there was no other method, at least at the time this was written

File: src/blockchain.js (hunk 6 above)
Comment (iurimatias) [reply src/blockchain.js:56]: they are for different things

**Review by jrainville — APPROVED** [review]

**Pull request merged**